plugins/woocommerce-admin/client/products/hooks/use-product-auto-save.ts

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import { debounce, pick } from 'lodash';
+import { Product } from '@woocommerce/data';
+import { useDebounce } from '@wordpress/compose';
+import { useEffect, useMemo, useState } from '@wordpress/element';
+import { useFormContext } from '@woocommerce/components';
+
+/**
+ * Internal dependencies
+ */
+import { useProductHelper } from '../use-product-helper';
+
+export const useProductAutoSave = ( dependencies: ( keyof Product )[] ) => {
+	const [ isAutoSaving, setIsAutoSaving ] = useState( false );
+	const { isValidForm, setValue, values } = useFormContext< Product >();
+	const { createOrUpdateProductWithStatus } = useProductHelper();
+
+	const dependencyValues = dependencies.reduce(
+		( acc: unknown[], key: keyof Product ) => {
+			acc.push( values[ key ] );
+			return acc;
+		},
+		[]
+	);
+
+	const handleProductUpdated = ( updatedProduct: Product ) => {
+		if ( ! updatedProduct || ! updatedProduct.id ) {
+			return;
+		}
+
+		setValue( 'id', updatedProduct.id );
```

Comment:
This triggers an update to the form somewhere upstream (presumably after the product is fetched) and then resets the form.

This can cause weird behavior if continuing to type after the debounce and resets the value.

My preference is to save this for a follow-up since its (kind of) a pre-existing issue with any updates being made during save, but I can tackle here if its quick and we know where this is stemming from.

Reply:
Hmm...I hit this when typing just a bit slow. I'm okay with addressing this in a follow up as long as we prioritize it, since this could be frustrating to users if they feel like their keystrokes are getting lost.

Reply:
Definitely.  Very frustrating and I've also hit this issue when typing slow.

I'm happy to address here if we think we can't get this done by next code freeze, but I'm slightly leaning towards a follow-up as I think we should be able to.